### PR TITLE
WIP: Feature: Subtracted Type

### PR DESCRIPTION
**Pull request by BackEndTea**
This is a first shot at the 'subtracted' type.

For now please ignore any style issues, missing features etc.

Fixes: https://github.com/phpstan/phpstan/issues/3182

Besides the new doc comments, this should also find issues like the following: https://phpstan.org/r/7356d0fd-79ef-4457-981e-d35f27077747

### CODE REVIEW

File: src/Type/SubtractedType.php [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+		/*
+		 * TODO:
+		 * Both the original and the subtracted type can access properties
+		 * now we need to determine if *only* the subtracted type has it, or if it is
+		 * present without the subtracted type as well.
+		 */
+		return TrinaryLogic::createMaybe();
+	}
+
+	public function hasProperty(string $propertyName): TrinaryLogic
```
Comment (BackEndTea) [inline src/Type/SubtractedType.php:111]: What do you think of this logic?

If both have the property we return a maybe, as we can't be sure where it came from.

Note that at this point i'm not 100% sure if it is even possible to subtract one class from another if we would 'normalize' union types during creation.

File: src/Type/SubtractedType.php [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+		$this->subtractedType = $subtractedType;
+	}
+
+	/**
+	 * TODO: move/use logic from TypeCombinator here, to create the type
+	 * @see TypeCombinator::remove()
+	 * The goal should be to normalize the type if possible (subtract a class from a union type of classes etc)
+	 * Only if that is not possible use this type, e.g. when subtracting a constant string from the general string type
+	 */
+	public static function create(Type $originalType, Type $subtractedType): Type
```
Comment (BackEndTea) [inline src/Type/SubtractedType.php:30]: See TODO, this type should only be used as a last resort, if we can subtract the types easily e.g. remove one type from a union or intersection then that should be done instead.

File: src/Type/SubtractedType.php [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+	}
+
+	public function getProperty(string $propertyName, ClassMemberAccessAnswerer $scope): PropertyReflection
+	{
+		/*
+		 * TODO: possible validate this
+		 * Assume for now that this is only called if the property is present, thus we can retrieve
+		 * it from the original type.
+		 */
+		return $this->originalType->getProperty($propertyName, $scope);
```
Comment (BackEndTea) [inline src/Type/SubtractedType.php:146]: Is it okay to just retrieve the original property? Prehaps we could check here again, and throw a shouldNotHappenException here.

**Comment by BackEndTea** [comment]
@ondrejmirtes I'd love some early feedback on this

**Comment by BackEndTea** [comment]
Also pining @JanTvrdik, i'd love to hear opinions/ideas for this.

**Comment by ondrejmirtes** [comment]
Haven't looked into it yet in more detail, but my idea is the same. The new Type implementation has to correctly answer to `isSuperTypeOf` and correctly work with `TypeCombinator::union()` and `TypeCombinator::intersect()`. That's all what I have to say about it :)

What's currently the biggest hurdle and will probably require one more phase before this one, is that a lot of code out there currently asks about types using `instanceof *Type`, like `instanceof ObjectType`. This `SubtractedType` will break that code.

What I'd like to do is to replace these use cases (for example you're asking about `instanceof ObjectType` because you want to ask `$type->getClassName()`) by adding methods to the `Type` interface one by one over time. And in the end, I'd like to disallow `instanceof *Type` with a PHPStan rule, for the next major version, if we're successful.

This work would also allow us to do more cool stuff - right now, there's `TemplateMixedType`, `TemplateObjectType`, `TemplateObjectWithoutClassType` so that `instanceof *Type` still works. Some `@template T of X` aren't supported because of this, for example where `X` is a scalar or a union.

WDYT?

**Comment by muglug** [comment]
You might be interested in Psalm's approach, where docblock assertions are not types – they're user-representations of the logical clauses the tool uses to model conditionals.

All conditionals in Psalm are first converted to clauses before being translated to types. This adds an intermediate step, but it makes it possible to avoid false-positives in code like this (abstracted from Vimeo's codebase).

```php
class RangeComponent {
    public function addSibling(RangeComponent $a) : void {}
    public function someMethod() : void {}
}

function formatRange(?RangeComponent $from, ?RangeComponent $to): void {
    if (!$to && !$from) {
        $to = new RangeComponent();
        $from = new RangeComponent();
    } elseif (!$from) {
        $from = new RangeComponent();
        $from->addSibling($to);
    } elseif (!$to) {
        $to = new RangeComponent();
        $to->addSibling($from);
    }
    $from->someMethod();
    $to->someMethod();
}
```

The `notNull` assertion [here](https://github.com/webmozart/assert/blob/e692710ce783cf3d42e7e157e7bb18c12b2107ec/src/Assert.php#L771) is converted into a `Clause` object in Psalm. At every conditional branch all the current clauses [are mined for things that can be asserted](https://github.com/vimeo/psalm/blob/20bcfb096a90a0a1778bf3754083a7d58115675c/src/Psalm/Type/Algebra.php#L477) about the current variables. Sometimes we can't say something definitive about a variable after one conditional, but can say something definitive in a child:

```
function formatRange(?RangeComponent $from, ?RangeComponent $to): void {
    if ($to || $from) {
        // the clauses generated above cannot be convered into types
        if (!$from) {
            // now they can!
            // $from is null, $to is non-null
        }
    }
}
```

While this might seem like a lot of work, it brings many benefits in terms of correctness. And, more importantly, it's quite a lot of fun to implement.

This demonstrates how `@psalm-assert` annotations can be used to stand-in for regular PHP ones: https://psalm.dev/r/a15c0be02a

**Comment by ondrejmirtes** [comment]
Types in PHPStan are converted to other types after assertions, this is baked into the project and it would be a gargantuan task to steer the ship elsewhere. We use intersection type math to achieve this and it works fine.

**Comment by ondrejmirtes** [comment]
There are some interesting "accessory" types like HasMethodType or NonEmptyArrayType that are supposed to be always used in intersections. Maybe this subtracting could be achieved with a similar accessory type? WDYT @JanTvrdik ?

**Comment by ondrejmirtes** [comment]
It would also break `instanceof *Type` but maybe it would be easier to implement than `new SubtractedType($a, $b)`.

**Comment by ondrejmirtes** [comment]
It's not time to do this now, in order to work properly, we first need to make `instanceof *Type` illegal by identifying the use-cases and moving them to methods on `Type`.

**Pull request closed without merging**